profiler/src/ProfilerEngine/Datadog.Profiler.Native.Windows/EtwEventsManager.cpp

```diff
@@ -0,0 +1,396 @@
+    IContentionListener* pContentionListener,
+    IGCSuspensionsListener* pGCSuspensionsListener,
+    IConfiguration* pConfiguration)
+    :
+    _pAllocationListener(pAllocationListener),
+    _pContentionListener(pContentionListener)
+{
+    _isDebugLogEnabled = pConfiguration->IsDebugLogEnabled();
+
+    _threadsInfo.reserve(256);
+    _parser = std::make_unique<ClrEventsParser>(
+        nullptr,  // to avoid duplicates with what is done in EtwEventsHandler
+        nullptr,  // to avoid duplicates with what is done in EtwEventsHandler
+        pGCSuspensionsListener);
+    _logger = std::make_unique<ProfilerLogger>();
+}
+
+
+uint64_t TimestampToEpochNS(uint64_t eventTimestamp)
+{
+    // the event timestamp is in 100ns units since 1601-01-01 in UTC
+    FILETIME ft;
+    ft.dwLowDateTime = eventTimestamp & 0xFFFFFFFF;
+    ft.dwHighDateTime = eventTimestamp >> 32;
+
+    // convert into system time
+    SYSTEMTIME st;
+    ::FileTimeToSystemTime(&ft, &st);
+
+    // take GMT shift into account
+    ::SystemTimeToTzSpecificLocalTime(nullptr, &st, &st);
+
+    // convert in epoch time
+    tm t = {0};
+    t.tm_year = st.wYear - 1900;
+    t.tm_mon = st.wMonth - 1;
+    t.tm_mday = st.wDay;
+    t.tm_hour = st.wHour;
+    t.tm_min = st.wMinute;
+    t.tm_sec = st.wSecond;
+    time_t timeSinceEpoch = mktime(&t);
+
+    return timeSinceEpoch * 1000'000'000 + st.wMilliseconds * 1'000'000;  // don't loose the milliseconds accuracy
+}
+
+void EtwEventsManager::OnEvent(
+    uint64_t systemTimestamp,
+    uint32_t tid,
+    uint32_t version,
+    uint64_t keyword,
+    uint8_t level,
+    uint32_t id,
+    uint32_t cbEventData,
+    const uint8_t* pEventData)
+{
+    if (keyword == KEYWORD_STACKWALK)
+    {
+        if (_isDebugLogEnabled)
+        {
+            //std::cout << "StackWalk" << std::endl;
+        }
+
+        auto pThreadInfo = GetOrCreate(tid);
+        if (pThreadInfo->LastEventWasContentionStart())
+        {
+            AttachContentionCallstack(pThreadInfo, cbEventData, pEventData);
+        }
+        else if (pThreadInfo->LastEventWasAllocationTick())
+        {
+            AttachAllocationCallstack(pThreadInfo, cbEventData, pEventData);
+        }
+
+        pThreadInfo->ClearLastEventId();
+    }
+    else if (keyword == KEYWORD_CONTENTION)
+    {
+        if (_pContentionListener == nullptr)
+        {
+            return;
+        }
+
+        if (id == EVENT_CONTENTION_START)
+        {
+            if (_isDebugLogEnabled)
+            {
+                std::cout << "ContentionStart" << std::endl;
+            }
+
+            auto pThreadInfo = GetOrCreate(tid);
+            pThreadInfo->ContentionStartTimestamp = systemTimestamp;
+            pThreadInfo->LastEventId = EventId::ContentionStart;
+        }
+        else if (id == EVENT_CONTENTION_STOP)
+        {
+            if (_isDebugLogEnabled)
+            {
+                std::cout << "ContentionStop" << std::endl;
```

Comment:
use the logger no?

Reply:
I'm using the logger for error/warning only
The _isDebugLogEnabled is for tracing to the console during tests and debugging